Source/Core/Core/AchievementManager.cpp

```diff
@@ -239,6 +249,38 @@ void AchievementManager::DoFrame()
   }
 }
 
+bool AchievementManager::CanPause()
+{
+  u32 frames_to_next_pause = 0;
+  bool can_pause = rc_client_can_pause(m_client, &frames_to_next_pause);
+  if (!can_pause)
+  {
+    OSD::AddMessage("Cannot spam pausing in hardcore mode.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    OSD::AddMessage(
+        fmt::format("Can pause in {} seconds.", static_cast<float>(frames_to_next_pause) / 60),
+        OSD::Duration::VERY_LONG, OSD::Color::RED);
+  }
+  return can_pause;
+}
+
+void AchievementManager::DoIdle()
+{
+  m_idle_thread = std::make_unique<std::thread>([]() {
+    while (Core::GetState(*AchievementManager::GetInstance().m_system) == Core::State::Paused &&
+           AchievementManager::GetInstance().m_background_execution_allowed)
```

Comment:
The way you're using m_background_execution_allowed, it technically needs to be an atomic to avoid undefined behavior.